### PR TITLE
Armv7 NEON and VFP for rzil

### PR DESCRIPTION
**Pull request by Heersin**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**
for early review

- NEON:
1. [x] VMOV
2. [x] VLDR/VSTR
3. [x] logical
4. [x] compare 
5. [x] vldn/vstn 
6. [ ] general data processing instruction (working on)

- VFP
1. [x] shared instructions
2. [ ] vcmp

**Test plan**
see #3282

**Closing issues**

### CODE REVIEW

File: librz/include/rz_il/rz_il_opbuilder_begin.h [hunk 1]
```diff
@@ -86,6 +86,10 @@
 #define FPOW(rmode, flx, fly)      rz_il_op_new_fpow(rmode, flx, fly)
 #define FMAD(rmode, flx, fly, flz) rz_il_op_new_fmad(rmode, flx, fly, flz)
 
+// TODO: add `feq` as prime operator in fbasic
+// https://smtlib.cs.uiowa.edu/theories-FloatingPoint.shtml
+#define FEQ(flx, fly) AND(FORDER(flx, fly), FORDER(fly, flx))
```
Comment (XVilka) [inline librz/include/rz_il/rz_il_opbuilder_begin.h]: @brightprogrammer I remember you asked about that

File: librz/analysis/arch/arm/arm_il32.c [hunk 2]
```diff
@@ -2626,8 +3253,8 @@ static RzILOpEffect *il_unconditional(csh *handle, cs_insn *insn, bool is_thumb)
 	case ARM_INS_TBH:
 		return tbb(insn, is_thumb);
 
-	// --
-	// Advanced SIMD and Floating-point
+		// --
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il32.c]: Unnecessary change

File: librz/analysis/arch/arm/arm_il32.c [hunk 3]
```diff
@@ -2361,7 +2988,7 @@ static RzILOpEffect *il_unconditional(csh *handle, cs_insn *insn, bool is_thumb)
 	case ARM_INS_PLDW:
 	case ARM_INS_PLI:
 	case ARM_INS_YIELD:
-	// barriers/synchronization
+		// barriers/synchronization
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il32.c:3998]: Same, unnecessary change

File: librz/analysis/arch/arm/arm_il32.c [hunk 4]
```diff
@@ -28,54 +29,102 @@ static const char *regs_bound_32[] = {
...
-	case ARM_REG_R5: return "r5";
-	case ARM_REG_R6: return "r6";
-	case ARM_REG_R7: return "r7";
-	case ARM_REG_R8: return "r8";
-	case ARM_REG_R9: return "r9";
-	case ARM_REG_R10: return "r10";
-	case ARM_REG_R11: return "r11";
-	case ARM_REG_R12: return "r12";
-	default: return NULL;
+	case ARM_REG_LR:
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il32.c]: It was ok condensed before, IMHO

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/include/rz_il/rz_il_opbuilder_begin.h (hunk 1 above)
Comment (brightprogrammer) [reply librz/include/rz_il/rz_il_opbuilder_begin.h]: Yes I did, there was no equality check for floating pt numbers.

File: librz/analysis/arch/arm/arm_il32.c [hunk 5]
```diff
@@ -252,6 +406,67 @@ static RzILOpBitVector *arg_mem(RzILOpBitVector *base_plus_disp, cs_arm_op *op,
...
+	ut64 tmp = imm;
+
+	if (dreg_width == 128) {
+		// for <Qd> registers
+		ut64 imm_low = 0;
+		ut64 imm_high = 0;
+
+		for (int i = 0; i < repeat_times / 2; ++i) {
+			imm_low += tmp;
+			imm_high += tmp;
```
Comment (brightprogrammer) [inline librz/analysis/arch/arm/arm_il32.c:430]: shouldn't these be logical OR instead of ADD operations?

File: librz/analysis/arch/arm/arm_il32.c [hunk 6]
```diff
@@ -252,6 +406,67 @@ static RzILOpBitVector *arg_mem(RzILOpBitVector *base_plus_disp, cs_arm_op *op,
...
+			tmp <<= imm_width;
+		}
+
+		return (APPEND(UN(64, imm_high), UN(64, imm_low)));
+	}
+
+	// for <Dd> and <Sd>
+	for (int i = 0; i < repeat_times; ++i) {
+		final_imm += tmp;
```
Comment (brightprogrammer) [inline librz/analysis/arch/arm/arm_il32.c:440]: same, logical OR

File: librz/analysis/arch/arm/arm_il32.c (hunk 4 above)
Comment (Heersin) [reply librz/analysis/arch/arm/arm_il32.c]: oh they were reformatted by the local `clang-format`

File: librz/include/rz_il/rz_il_opbuilder_begin.h (hunk 1 above)
Comment (Heersin) [inline librz/include/rz_il/rz_il_opbuilder_begin.h]: #define FEQ(flx, fly) INV(AND(FORDER(flx, fly), FORDER(fly, flx)))

File: librz/include/rz_il/rz_il_opbuilder_end.h [hunk 7]
```diff
@@ -127,4 +127,38 @@
...
+#undef FCONVERT_TO
+#undef FORDER
+#undef FROUND
+#undef FSQRT
+#undef FADD
+#undef FSUB
+#undef FMUL
+#undef FDIV
+#undef FMOD
+#undef FMAD
```
Comment (Heersin) [inline librz/include/rz_il/rz_il_opbuilder_end.h]: Remove those out-dated `#undef`

File: librz/analysis/arch/arm/arm_il32.c (hunk 5 above)
Comment (Heersin) [reply librz/analysis/arch/arm/arm_il32.c:430]: it could be discussed later. I think a repeated immediate value would be much clear than immediate value with multiple OR. Though logical OR of the immediate value seems more precise from semantic perspective.

File: librz/analysis/arch/arm/arm_il32.c [hunk 8]
```diff
@@ -2314,22 +2515,880 @@ static RzILOpEffect *tbb(cs_insn *insn, bool is_thumb) {
...
+	ut32 elem_bytes = elem_bits / 8;
+	ut32 lanes = 64 / elem_bits;
+
+	RzILOpEffect *wback_eff = NULL;
+	RzILOpEffect *eff = NULL;
+	RzILOpBitVector *addr = ARG(rn_idx);
+
+	for (int i = 0; i < n_groups; ++i) {
+		for (int j = 0; j < lanes; ++j) {
+			RzILOpBitVector *data0, *data1, *data2, *data3;
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il32.c:3013]: You can extract body of these loops into a separate function to improve readability

File: librz/analysis/arch/arm/arm_il32.c [hunk 9]
```diff
@@ -2314,22 +2543,420 @@
 	return JMP(ADD(U32(PC(insn->address, is_thumb)), SHIFTL0(UNSIGNED(32, off), UN(5, 1))));
 }
 
+static RzILOpEffect *write_reg_lane(arm_reg reg, ut32 lane, ut32 vec_size, RzILOpBitVector *v) {
+	ut32 reg_size = reg_bits(reg);
+	RzILOpBitVector *sft_val = SHIFTL0(UNSIGNED(reg_size, v), UN(8, vec_size * lane));
```
[the remark this answers is not in the document]
Comment (XVilka) [reply librz/analysis/arch/arm/arm_il32.c]: Notice this CodeQL warning

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/analysis/arch/arm/arm_il32.c [hunk 10]
```diff
@@ -2314,22 +2515,880 @@ static RzILOpEffect *tbb(cs_insn *insn, bool is_thumb) {
...
+		return NULL;
+	}
+
+	// to single lane
+	if (NEON_LANE(0) != -1) {
+		return vldn_single_lane(insn, is_thumb);
+	}
+
+	// capstone cannot distinguish details of the following instructions
+	// vld3.8 {d0, d1, d2}, [r0] (f420040f)
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il32.c:3259]: This one should be addressed by https://github.com/capstone-engine/capstone/pull/1949
Please add `TODO` or `FIXME` to update once we merge our `capstone-ng` branch. cc @Rot127

**Review by XVilka — CHANGES_REQUESTED** [review]
Please also add tests for those

File: librz/analysis/arch/arm/arm_il32.c [hunk 11]
```diff
@@ -101,15 +137,75 @@ static RzILOpBitVector *read_reg(ut64 pc, arm_reg reg) {
 		RzILOpBitVector *var = VARG(reg_var_name(ARM_REG_D0 + idx / 2));
 		return UNSIGNED(32, idx % 2 ? SHIFTR0(var, UN(7, 32)) : var);
 	}
+	if (reg >= ARM_REG_Q0 && reg <= ARM_REG_Q15) {
+		ut32 low_dr_idx = (reg - ARM_REG_Q0) << 1;
```
Comment (Rot127) [inline librz/analysis/arch/arm/arm_il32.c:141]: ~~This is risky, because this assumes that the enum order never changes. But it might do (and will with the update). Please calculate the `DR` index by getting the `Qx` offset from `Q0` first and then add it to `D0` accordingly.~~
Wait. I am confused. It should be `ARM_REG_D0 + ((reg - ARM_REG_Q0) << 1)`. Right?

File: librz/analysis/arch/arm/arm_il32.c [hunk 12]
```diff
@@ -101,15 +137,75 @@ static RzILOpBitVector *read_reg(ut64 pc, arm_reg reg) {
...
 	return var ? VARG(var) : NULL;
 }
 
+/**
+ * Return IL of bitvector store in register lane
+ * The length of such bitv is `data_size`
+ */
+static RzILOpBitVector *read_reg_lane(arm_reg reg, ut32 lane, ut32 data_size) {
+	if (is_core_reg(reg)) {
+		return NULL;
```
Comment (Rot127) [inline librz/analysis/arch/arm/arm_il32.c:159]: Is this a place for `rz_warn_if_reached();`?

File: librz/analysis/arch/arm/arm_il32.c [hunk 13]
```diff
@@ -101,15 +137,75 @@ static RzILOpBitVector *read_reg(ut64 pc, arm_reg reg) {
...
+	case ARM_VECTORDATA_S16:
+	case ARM_VECTORDATA_U16:
+		return 16;
+	case ARM_VECTORDATA_I64:
+	case ARM_VECTORDATA_F64:
+	case ARM_VECTORDATA_U64:
+	case ARM_VECTORDATA_S64:
+		return 64;
+	case ARM_VECTORDATA_INVALID:
+		return 0;
```
Comment (Rot127) [inline librz/analysis/arch/arm/arm_il32.c]: `rz_warn_if_reached();` as well?

File: librz/analysis/arch/arm/arm_il32.c [hunk 14]
```diff
@@ -252,6 +358,87 @@ static RzILOpBitVector *arg_mem(RzILOpBitVector *base_plus_disp, cs_arm_op *op,
...
 }
 
+/**
+ * Replicate given value to `dreg_width` length
+ * Note the ownership of `val` will be transfered
+ */
+static RzILOpBitVector *replicated_val(ut32 val_width, ut32 dreg_width, RzILOpBitVector *val) {
+	ut32 repeat_times = dreg_width / val_width;
+	if (dreg_width % val_width != 0) {
+		return NULL;
```
Comment (Rot127) [inline librz/analysis/arch/arm/arm_il32.c:401]: `rz_warn_if_reached()`?

File: librz/analysis/arch/arm/arm_il32.c [hunk 15]
```diff
@@ -252,6 +358,87 @@ static RzILOpBitVector *arg_mem(RzILOpBitVector *base_plus_disp, cs_arm_op *op,
 	return base_plus_disp;
 }
 
+/**
+ * Replicate given value to `dreg_width` length
+ * Note the ownership of `val` will be transfered
+ */
+static RzILOpBitVector *replicated_val(ut32 val_width, ut32 dreg_width, RzILOpBitVector *val) {
```
Comment (Rot127) [inline librz/analysis/arch/arm/arm_il32.c]:
```suggestion
/**
 * Replicate given value to `dreg_width` length
 */
static RzILOpBitVector *replicated_val(ut32 val_width, ut32 dreg_width, RZ_OWN RzILOpBitVector *val) {
```

File: librz/analysis/arch/arm/arm_il32.c [hunk 16]
```diff
@@ -252,6 +358,87 @@ static RzILOpBitVector *arg_mem(RzILOpBitVector *base_plus_disp, cs_arm_op *op,
...
+	final_imm = tmp;
+	for (int i = 0; i < repeat_times - 1; ++i) {
+		final_imm += tmp;
+		tmp <<= imm_width;
+	}
+
+	return UN(dreg_width, final_imm);
+}
+
+static ut32 get_imm(cs_insn *insn, int n, RZ_NULLABLE RzILOpBool **carry_out) {
```
Comment (Rot127) [inline librz/analysis/arch/arm/arm_il32.c:454]: Docs please. Also describing the arguments.

File: librz/analysis/arch/arm/arm_il32.c [hunk 17]
```diff
@@ -482,7 +652,9 @@ static RzILOpEffect *add_sub(cs_insn *insn, bool is_thumb) {
 		// alias for adr
 		return adr(insn, is_thumb);
 	}
-	bool is_sub = insn->id == ARM_INS_SUB || insn->id == ARM_INS_SUBW || insn->id == ARM_INS_RSB || insn->id == ARM_INS_RSC || insn->id == ARM_INS_SBC;
+	bool is_sub =
```
Comment (Rot127) [inline librz/analysis/arch/arm/arm_il32.c:695]: Maybe move those long ORs into their own inline function. Looks a little nicer. Also above and below.

File: librz/analysis/arch/arm/arm_il32.c [hunk 18]
```diff
@@ -2314,22 +2515,880 @@ static RzILOpEffect *tbb(cs_insn *insn, bool is_thumb) {
...
+		return SEQ2(write_reg(REGID(0), rt1_val),
+			write_reg(REGID(1), rt2_val));
+	}
+
+	// core register to scalar
+	if (NEON_LANE(0) != -1 && !is_core_reg(REGID(0)) && is_core_reg(REGID(1))) {
+		// vmov.<vec_size> <Dd>[lane], <Rt>
+		// <Dd>[lane] = <Rt>{vecsize - 1 : 0}
+		if (!VVEC_SIZE(insn) || NEON_LANE(0) == -1) {
+			return NULL;
```
Comment (Rot127) [inline librz/analysis/arch/arm/arm_il32.c:2657]: `rz_warn_if_reached()`?

Comment (Rot127) [reply librz/analysis/arch/arm/arm_il32.c:2657]: Same below. Basically whenever `NULL` is returned and this means something unexpected happened.

File: librz/analysis/arch/arm/arm_il32.c [hunk 19]
```diff
@@ -2314,22 +2515,880 @@ static RzILOpEffect *tbb(cs_insn *insn, bool is_thumb) {
...
+	}
+
+	// if <Rt> is APSR, transfer to flags
+	if (REGID(1) == ARM_REG_APSR_NZCV) {
+		RzILOpBitVector *val = VARG("fpscr");
+		return SEQ4(
+			SETG("nf", INV(IS_ZERO(LOGAND(val, U32(1ul << 31))))),
+			SETG("zf", INV(IS_ZERO(LOGAND(DUP(val), U32(1ul << 30))))),
+			SETG("cf", INV(IS_ZERO(LOGAND(DUP(val), U32(1ul << 29))))),
+			SETG("vf", INV(IS_ZERO(LOGAND(DUP(val), U32(1ul << 28))))));
```
Comment (Rot127) [inline librz/analysis/arch/arm/arm_il32.c:2717]: It might be worth to move this into its own `set_flags` function. Because it repeats below.

File: librz/analysis/arch/arm/arm_il32.c [hunk 20]
```diff
@@ -2314,22 +2515,880 @@ static RzILOpEffect *tbb(cs_insn *insn, bool is_thumb) {
...
+		// vldn {list}, [Rn]
+		rm_idx = -1;
+		regs = OPCOUNT() - 1;
+	}
+
+	// mem_idx
+	rn_idx = regs;
+
+	// assert list_size % n == 0
+	// assert they were all Dn
```
Comment (Rot127) [inline librz/analysis/arch/arm/arm_il32.c:3001]: Those asserts can be checked with `rz_warn_if...` functions. They get removed in the Release build

File: librz/analysis/arch/arm/arm_il32.c [hunk 21]
```diff
@@ -2314,22 +2515,880 @@ static RzILOpEffect *tbb(cs_insn *insn, bool is_thumb) {
...
+		wback_eff = write_reg(rn_idx, ADD(REG_VAL(rn_idx), new_offset));
+	} else {
+		wback_eff = NOP();
+	}
+
+	return SEQ2(eff, wback_eff);
+}
+
+static RzILOpEffect *vldn(cs_insn *insn, bool is_thumb) {
+	if (OPCOUNT() < 2 || !ISREG(0)) {
```
Comment (Rot127) [inline librz/analysis/arch/arm/arm_il32.c:3249]: `rz_return_if_fail()` if this is an error case.

File: librz/analysis/arch/arm/arm_il32.c (hunk 10 above)
Comment (Rot127) [reply librz/analysis/arch/arm/arm_il32.c:3259]: Unfortunately not. Capstone assigns instruction IDs by their Mnemonic (which is nice, because we do not have to deal with the original >4000 IDs). But this means that both of them are classified as `ARM_INS_VLD3`.

The LLVM id is not accessible from the CS API. We need to extend the API for such cases. It is related to https://github.com/capstone-engine/capstone/issues/1945. Access to the raw TableGen info is requested there as well.

File: librz/analysis/arch/arm/arm_il32.c [hunk 22]
```diff
@@ -2314,22 +2515,880 @@ static RzILOpEffect *tbb(cs_insn *insn, bool is_thumb) {
...
+		return NULL;
+	}
+
+	bool wback = insn->detail->arm.writeback;
+	RzILOpEffect *wback_eff;
+	if (wback) {
+		RzILOpBitVector *new_offset = use_rm ? ARG(rm_idx) : UN(32, elem_bytes * n);
+		wback_eff = write_reg(rn_idx, ADD(REG_VAL(rn_idx), new_offset));
+	} else {
+		wback_eff = NOP();
```
Comment (Rot127) [inline librz/analysis/arch/arm/arm_il32.c]: Better use `EMPTY()` instead of `NOP()`

**Review by Rot127 — COMMENTED** [review]
Nice work!
Please make sure to add the `rz_warn_if_reached()` and other asserts wherever something unexpected happens. It will be of immeasurable help when I need to rebase this on the `capstone-ng`.

File: librz/analysis/arch/arm/arm_il32.c (hunk 11 above)
Comment (Heersin) [reply librz/analysis/arch/arm/arm_il32.c:141]: oh,  yes. My mistake. Thank you Rot127

File: librz/analysis/arch/arm/arm_il32.c (hunk 22 above)
Comment (Heersin) [reply librz/analysis/arch/arm/arm_il32.c]: Yes

File: librz/analysis/arch/arm/arm_il32.c (hunk 19 above)
Comment (Heersin) [reply librz/analysis/arch/arm/arm_il32.c:2717]: not really, the below one is read value from those flags. A similar function is "update_flags_from_cpsr", but we only focus on 4 flag bits instead the whole `cpsr`. So I think it's ok to keep it.

**Comment by XVilka** [comment]
@Heersin do you have any updates for this PR?

File: librz/analysis/arch/arm/arm_il32.c [hunk 23]
```diff
@@ -155,7 +155,8 @@ static RzILOpBitVector *read_reg(ut64 pc, arm_reg reg) {
  */
 static RzILOpBitVector *read_reg_lane(arm_reg reg, ut32 lane, ut32 data_size) {
 	if (is_core_reg(reg)) {
-		return NULL;
+        rz_warn_if_reached();
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il32.c]: broken indentation

File: librz/analysis/arch/arm/arm_il32.c (hunk 14 above)
Comment (XVilka) [reply librz/analysis/arch/arm/arm_il32.c:401]: You forgot this one @Heersin. Please also rebase.

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by XVilka** [comment]
@Heersin @brightprogrammer note, NEON has `VCVT` instruction that converts between fixed point and floating point (as one of the possible scenarios), and it's presented in the `float_ex1` testcase:
```
0000070c  bdeee78a   vcvt.s32.f32 s16, s15
```


VCVT | Convert fixed-point or integer to floating point, floating-point to integer or fixed-point
-- | --
VCVT | Convert between half-precision and single-precision floating-point numbers

File: librz/analysis/arch/arm/arm_il32.c (hunk 14 above)
Comment (Heersin) [reply librz/analysis/arch/arm/arm_il32.c:401]: ok. btw rebase and squash some commits ?

**Comment by brightprogrammer** [comment]
> @Heersin @brightprogrammer note, NEON has `VCVT` instruction that converts between fixed point and floating point (as one of the possible scenarios), and it's presented in the `float_ex1` testcase:
> 
> ```
> 0000070c  bdeee78a   vcvt.s32.f32 s16, s15
> ```
> 
> VCVT	Convert fixed-point or integer to floating point, floating-point to integer or fixed-point
> VCVT	Convert between half-precision and single-precision floating-point numbers

yes, for MIPS too
`   120000da0:	46000021 	cvt.d.s	$f0,$f0`
to covert floating or fixed point to double floating point.

Will we have support for fixed point conversions in near future?

**Comment by XVilka** [comment]
> > @Heersin @brightprogrammer note, NEON has `VCVT` instruction that converts between fixed point and floating point (as one of the possible scenarios), and it's presented in the `float_ex1` testcase:
> > ```
> > 0000070c  bdeee78a   vcvt.s32.f32 s16, s15
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > VCVT	Convert fixed-point or integer to floating point, floating-point to integer or fixed-point
> > VCVT	Convert between half-precision and single-precision floating-point numbers
> 
> yes, for MIPS too ` 120000da0: 46000021 cvt.d.s $f0,$f0` to covert floating or fixed point to double floating point.
> 
> Will we have support for fixed point conversions in near future?

I think we can emulate one of the following algorithms in our IL: 
- https://embeddedartistry.com/blog/2018/07/12/simple-fixed-point-conversion-in-c/
- https://stackoverflow.com/questions/4368396/how-do-i-convert-a-floating-point-c-code-to-fixed-point

It can become even some common macro

**Comment by XVilka** [comment]
@Heersin please focus on the instructions that are presented in that test file, so that it could be finished and tested early, then merged. The rest of instructions could be implemented later and in separate PRs. MVP is the most optimal strategy in this case.

**Comment by Heersin** [comment]
> @Heersin please focus on the instructions that are presented in that test file, so that it could be finished and tested early, then merged. The rest of instructions could be implemented later and in separate PRs. MVP is the most optimal strategy in this case.

sounds good, It's a nice strategy.

**Review by XVilka — CHANGES_REQUESTED** [review]
```
test_il_validate_pure_forder ERR
[XX] Fail at line 1464: report: expected Op forder formats of left operand (float:1) and right operand (float:0) do not agree., got Op forder formats of left operand (float:2) and right operand (float:1) do not agree..

test_il_validate_pure_float_binop_with_round ERR
[XX] Fail at line 1510: report: expected Op fadd formats of left operand (float:1) and right operand (float:0) do not agree., got Op fadd formats of left operand (float:2) and right operand (float:1) do not agree..

test_il_validate_pure_float_terop_with_round ERR
[XX] Fail at line 1570: report: expected types of operand in op fmad do not agree: operand1 (float:1) operand2 (float:0) operand3 (float:0), got types of operand in op fmad do not agree: operand1 (float:2) operand2 (float:1) operand3 (float:1).
```

File: librz/analysis/arch/arm/arm_il32.c [hunk 24]
```diff
@@ -101,15 +137,108 @@ static RzILOpBitVector *read_reg(ut64 pc, arm_reg reg) {
...
 }
 
+/**
+ * Return IL of bitvector store in register lane
+ * The length of such bitv is `data_size`
+ */
+static RzILOpBitVector *read_reg_lane(arm_reg reg, ut32 lane, ut32 data_size) {
+	if (is_core_reg(reg)) {
+		rz_warn_if_reached();
+		return NULL;
```
Comment (DMaroo) [inline librz/analysis/arch/arm/arm_il32.c:159]: Wouldn't it be better to use `rz_return_val_if_fail` here? Would be more concise.

**Review by DMaroo — CHANGES_REQUESTED** [review]

File: librz/analysis/arch/arm/arm_il32.c (hunk 24 above)
Comment (Heersin) [reply librz/analysis/arch/arm/arm_il32.c:159]: yep, would consider that

Comment (DMaroo) [reply librz/analysis/arch/arm/arm_il32.c:159]: It is generally nicer to have `rz_return_*` functions at the start of a function rather than `rz_warn_if_reached`. So try to replace it other such places as well. Thanks!

**Comment by XVilka** [comment]
For the record (even if it's AARCH64, still would be similar): https://discourse.llvm.org/t/armv8-fixed-point-to-floating-point-uses-division/71272

File: librz/analysis/arch/arm/arm_il32.c (hunk 17 above)
Comment (XVilka) [reply librz/analysis/arch/arm/arm_il32.c:695]: Agree on that

**Review by XVilka — CHANGES_REQUESTED** [review]
Please add also disassembly/RzIL tests for these instructions, and rebase.

File: librz/analysis/arch/arm/arm_il32.c (hunk 17 above)
Comment (Heersin) [reply librz/analysis/arch/arm/arm_il32.c:695]: ok

**Comment by XVilka** [comment]
Please rebase to resolve the conflict

**Review by XVilka — CHANGES_REQUESTED** [review]
Please update 4 tests with a line with FPSCR register:
```
[XX] db/abi/platforms/reg_profile MAIN_ABI : arm-linux_gnu_guneabi register
RZ_NOPLUGINS=1 /home/runner/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -A -Qc ar bins/abi_bins/elf/platforms/arm-linux-gnueabi-echo
-- stdout
--- expected
+++ actual
@@ -22,3 +22,4 @@
 r14 = 0x00000000
 r15 = 0x00000000
 cpsr = 0x00000000
+fpscr = 0x00000000

[XX] db/abi/platforms/reg_profile MAIN_ABI : arm-linux_android_abi register
RZ_NOPLUGINS=1 /home/runner/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -A -Qc ar bins/abi_bins/elf/platforms/arm-linux-androideabi-echo
-- stdout
--- expected
+++ actual
@@ -22,3 +22,4 @@
 r14 = 0x00000000
 r15 = 0x00000000
 cpsr = 0x00000000
+fpscr = 0x00000000

[XX] db/formats/elf/core core regs linux arm32
RZ_NOPLUGINS=1 /home/runner/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 'ar
' bins/elf/core/core-linux-arm32
-- stdout
--- expected
+++ actual
@@ -22,3 +22,4 @@
 r14 = 0x76e60718
 r15 = 0x00010420
 cpsr = 0x60000010
+fpscr = 0x00000000

 [XX] db/analysis/arm arm show registers in table
RZ_NOPLUGINS=1 /home/runner/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 'e asm.arch=arm
e asm.bits=32
ar=
' =
-- stdout
--- expected
+++ actual
@@ -4,3 +4,4 @@
      r5 0x00000000           r6 0x00000000           r7 0x00000000           r8 0x00000000
      r9 0x00000000          r10 0x00000000          r11 0x00000000          r12 0x00000000
     r13 0x00000000          r14 0x00000000          r15 0x00000000         cpsr 0x00000000
+  fpscr 0x00000000  
```

File: librz/analysis/arch/arm/arm_il32.c (hunk 14 above)
Comment (XVilka) [reply librz/analysis/arch/arm/arm_il32.c:401]: @Heersin yes, rebasing and squashing

File: librz/analysis/arch/arm/arm_il32.c (hunk 13 above)
Comment (XVilka) [reply librz/analysis/arch/arm/arm_il32.c]: @Heersin you forgot this one

File: librz/analysis/arch/arm/arm_il32.c [hunk 25]
```diff
@@ -252,6 +391,95 @@
...
+	ut32 repeat_times = dreg_width / val_width;
+	if (dreg_width % val_width != 0) {
+		rz_warn_if_reached();
+		return NULL;
+	}
+
+	RzILOpBitVector *ext_val = UNSIGNED(dreg_width, val);
+	RzILOpBitVector *rep_val = ext_val;
+	for (int i = 0; i < repeat_times - 1; ++i) {
+		rep_val = LOGOR(rep_val, SHIFTL0(DUP(ext_val), UN(8, val_width * i)));
```
[the remark this answers is not in the document]
Comment (XVilka) [reply librz/analysis/arch/arm/arm_il32.c:407]: Don't forget about CodeQL alerts as well

File: librz/analysis/arch/arm/arm_il32.c (hunk 17 above)
Comment (XVilka) [reply librz/analysis/arch/arm/arm_il32.c:695]: And this one

File: librz/analysis/arch/arm/arm_il32.c (hunk 18 above)
Comment (XVilka) [reply librz/analysis/arch/arm/arm_il32.c:2657]: And this one

File: librz/analysis/arch/arm/arm_il32.c [hunk 26]
```diff
@@ -2314,22 +2515,880 @@
...
+		break;
+	default:
+		rz_warn_if_reached();
+		return NULL;
+	}
+
+	bool wback = insn->detail->arm.writeback;
+	RzILOpEffect *wback_eff;
+	if (wback) {
+		RzILOpBitVector *new_offset = use_rm ? ARG(rm_idx) : UN(32, elem_bytes * n);
```
[the remark this answers is not in the document]
Comment (XVilka) [reply librz/analysis/arch/arm/arm_il32.c:3153]: @Heersin and this one

File: librz/analysis/arch/arm/arm_il32.c [hunk 27]
```diff
@@ -2314,22 +2515,880 @@
...
```
[the remark this answers is not in the document]
Comment (XVilka) [reply librz/analysis/arch/arm/arm_il32.c:3433]: And this one

**Comment by XVilka** [comment]
@Heersin did you check what instructions left to implement for this binary to be emulated successfully?

**Comment by XVilka** [comment]
There is also broken formatting:
```
Formatting [1341/1606] librz/include/rz_il/rz_il_opbuilder_begin.h
librz/include/rz_il/rz_il_opbuilder_begin.h:94:17: error: code should be clang-formatted [-Wclang-format-violations]
#define IL_FALSE  rz_il_op_new_b0()
                ^
librz/include/rz_il/rz_il_opbuilder_begin.h:95:16: error: code should be clang-formatted [-Wclang-format-violations]
#define IL_TRUE   rz_il_op_new_b1()
               ^
librz/include/rz_il/rz_il_opbuilder_begin.h:96:15: error: code should be clang-formatted [-Wclang-format-violations]
#define INV(x)    rz_il_op_new_bool_inv(x)
              ^
librz/include/rz_il/rz_il_opbuilder_begin.h:97:18: error: code should be clang-formatted [-Wclang-format-violations]
#define XOR(x, y) rz_il_op_new_bool_xor(x, y)
                 ^
librz/include/rz_il/rz_il_opbuilder_begin.h:98:18: error: code should be clang-formatted [-Wclang-format-violations]
#define AND(x, y) rz_il_op_new_bool_and(x, y)
                 ^
librz/include/rz_il/rz_il_opbuilder_begin.h:99:17: error: code should be clang-formatted [-Wclang-format-violations]
#define OR(x, y)  rz_il_op_new_bool_or(x, y)
                ^
```

**Comment by Heersin** [comment]
> 

all instruction in the main can be emualted. the left instruction in reported by `rz-tracetest` are `mcr, vmsr and vcmp` in the whole binary

btw is there a convenient command to emualte binary likes `qemu-arm binary` in rizin ? except the `aezsu` @XVilka

**Comment by XVilka** [comment]
> btw is there a convenient command to emualte binary likes `qemu-arm binary` in rizin ? except the `aezsu` @XVilka

No. What's wrong with `aezsu`? It's used in "emulateme" tests for RzIL already.

**Comment by Heersin** [comment]
> > btw is there a convenient command to emualte binary likes `qemu-arm binary` in rizin ? except the `aezsu` @XVilka
> 
> No. What's wrong with `aezsu`? It's used in "emulateme" tests for RzIL already.


no problem. just feel not that easy to use when compare to `qemu-[arch]` :)

File: librz/analysis/arch/arm/arm_il32.c (hunk 25 above)
[the remark this answers is not in the document]
Comment (Heersin) [reply librz/analysis/arch/arm/arm_il32.c:407]: could I dismiss them, since there are some constraints which permits they don't overflow. @XVilka

[the remark this answers is not in the document]
Comment (XVilka) [reply librz/analysis/arch/arm/arm_il32.c:407]: Yes. Once merged we will also be able to run Coverity on them.

File: test/db/rzil/arm32 [hunk 28]
```diff
@@ -268,3 +268,36 @@ EOF
...
+ar d7
+aezsu 0x10554 # after vmul
+ar d7
+aezs # after vadd
+ar d7
+aezs # after vcvt
+ar d16
+aezsu 0x1058c # before printf
+ar r2
+ar r3
```
Comment (XVilka) [inline test/db/rzil/arm32:292]: Please test all 3 arguments - a, b, and c

**Review by XVilka — CHANGES_REQUESTED** [review]

**Review by XVilka — CHANGES_REQUESTED** [review]
Also formatting problem:
```
 Formatting [1335/1606] librz/include/rz_il/rz_il_opbuilder_begin.h
librz/include/rz_il/rz_il_opbuilder_begin.h:94:17: error: code should be clang-formatted [-Wclang-format-violations]
#define IL_FALSE  rz_il_op_new_b0()
                ^
librz/include/rz_il/rz_il_opbuilder_begin.h:95:16: error: code should be clang-formatted [-Wclang-format-violations]
#define IL_TRUE   rz_il_op_new_b1()
               ^
librz/include/rz_il/rz_il_opbuilder_begin.h:96:15: error: code should be clang-formatted [-Wclang-format-violations]
#define INV(x)    rz_il_op_new_bool_inv(x)
              ^
librz/include/rz_il/rz_il_opbuilder_begin.h:97:18: error: code should be clang-formatted [-Wclang-format-violations]
#define XOR(x, y) rz_il_op_new_bool_xor(x, y)
                 ^
librz/include/rz_il/rz_il_opbuilder_begin.h:98:18: error: code should be clang-formatted [-Wclang-format-violations]
#define AND(x, y) rz_il_op_new_bool_and(x, y)
                 ^
librz/include/rz_il/rz_il_opbuilder_begin.h:99:17: error: code should be clang-formatted [-Wclang-format-violations]
#define OR(x, y)  rz_il_op_new_bool_or(x, y)
                ^
```

**Review by XVilka — APPROVED** [review]
